do not lock up addprocs on worker setup errors

tanmaykm opened the pull request:
`create_worker` (invoked during `addprocs`) waits on a message from the worker to indicate success. If the worker process terminates before sending this response, `create_worker` (and therefore `addprocs`) will remain locked up.

Usually the master process does become aware of a terminated worker, when the communication channel between them breaks due to worker exiting. The message processing loop exits as a result. But `create_worker` keeps on waiting though.

This commit introduces an additional task (timer) that monitors this message processing loop while master is waiting for a `JoinCompleteMsg` response from a worker. It makes `create_worker` return both when setup is successful (master receives a `JoinCompleteMsg`) and also when worker is terminated. Return value of `create_worker` is `0` when worker setup fails, instead of worker id when it is successful. Return value of `addprocs` contains only workers that were successfully launched and connected to. Added tests.

tanmaykm committed:
do not lock up addproc on worker setup errors

`create_worker` (invoked during `addprocs`) waits on a message from the worker to indicate success. If the worker process terminates before sending this response, `create_worker` (and therefore `addprocs`) will remain locked up.

Usually the master process does become aware of a terminated worker, when the communication channel between them breaks due to worker exiting. The message processing loop exits as a result.

This commit introduces an additional task (timer) that monitors this message processing loop while master is waiting for a `JoinCompleteMsg` response from a worker. It makes `create_worker` return both when setup is successful (master receives a `JoinCompleteMsg`) and also when worker is terminated. Return value of `create_worker` is 0 when worker setup fails, instead of worker id when it is successful. Return value of `addprocs` contains only workers that were successfully launched and connected to. Added some tests for that too.

amitmurthy commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -597,7 +600,11 @@ function create_worker(manager, wconfig)
-    process_messages(w.r_stream, w.w_stream, false)
+    procmsg_task = process_messages(w.r_stream, w.w_stream, false)
+    Timer(1, interval=1) do timer
+        istaskstarted(procmsg_task) && istaskdone(procmsg_task) && put!(rr_ntfy_join, nothing)
```
Can this be enhanced to also handle a handshake timeout case? i.e., the initial connect is successfully established, but the handshake does not complete within `Distributed.worker_timeout()` seconds (note that the worker process is not dead, just hanging in this scenario and so the master-worker connection is still up).

We can have `put!(rr_ntfy_join, :OK)`, `put!(rr_ntfy_join, :TIMEDOUT)` and `put!(rr_ntfy_join, :ERROR)` to differentiate between the different scenarios.

tanmaykm replied:
yes, `istaskstarted` is redundant. will remove

tanmaykm replied:
will also try handling the timeout case

tanmaykm replied:
@amitmurthy I've added a commit that handles the timeout case too: https://github.com/JuliaLang/julia/pull/32290/commits/2d2649cf7235dddb98dcdfcc6f3b9f43f757b82c

tanmaykm replied:
Note that the worker process may not be dead yet, and it may still be listed in `workers()` until it actually goes down.

tanmaykm commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -1554,7 +1554,7 @@ end
 # Issue # 22865
 # Must be run on a new cluster, i.e., all workers must be in the same state.
-rmprocs(workers())
+(nprocs() > 1) && rmprocs(workers())
```
throws a warning when there are no workers otherwise, which is confusing

tanmaykm committed:
do not lock up addproc on worker setup timeout

This is a corollary to the previous commit in #32290, and implements suggestions thereof.

It restricts the master to wait for a worker to respond within `Distributed.worker_timeout()` seconds. Beyond that it releases the lock on `rr_ntfy_join` with a special flag `:TIMEDOUT`. This flag is set to `:ERROR` in case of any errors during worker setup, and to `:OK` when the master received a `JoinCompleteMsg` indicating setup completion from worker.

`addprocs` returns the worker id in the list of workers it added only if it has received a `JoinCompleteMsg`, that is, only when `rr_ntfy_join` contains `:OK`. Note that the worker process may not be dead yet, and it may still be listed in `workers()` until it actually goes down.

amitmurthy commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -658,7 +671,7 @@ function create_worker(manager, wconfig)
     end
 
-    return w.id
+    return (istaskdone(procmsg_task) || (first(rr_ntfy_join.c.data) !== :OK)) ? 0 : w.id
```
Replace `first(rr_ntfy_join.c.data)` with `fetch(rr_ntfy_join.c)`. Avoids using internals of the `Channel` implementation.

Also the `istaskdone(procmsg_task)` check can be removed since we will come here only after `wait(rr_ntfy_join)` has been triggered.

tanmaykm replied:
Done.

amitmurthy commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -1592,10 +1592,39 @@ function reuseport_tests()
+    # test condition where connect times out
+    # using a non routable IP to test, ref: https://tools.ietf.org/html/rfc5737
+    Distributed.PGRP.workers[2].config.connect_at = ("203.0.113.0", w2_connect_at[2])
+    npids = addprocs_with_testenv(1; topology=:all_to_all, lazy=false)
```
Default worker timeout if ENV `"JULIA_WORKER_TIMEOUT"` is unset is 60.0 seconds. This will unnecessarily delay the test suite run. Better to test with a smaller value - see an example in https://github.com/JuliaLang/julia/blob/b7774095f62e47bb2c8467b8a328075ab5a3fb89/stdlib/Distributed/test/distributed_exec.jl#L1149 and 

https://github.com/JuliaLang/julia/blob/b7774095f62e47bb2c8467b8a328075ab5a3fb89/stdlib/Distributed/test/distributed_exec.jl#L1159-L1161

which sets JULIA_WORKER_TIMEOUT to 1 second for one of the testruns.

tanmaykm replied:
Done, `addprocs` times out quick now. But `rmprocs()` still takes a bit of time when the worker is stuck.

tanmaykm replied:
Not sure if issuing a `kill` to the stuck test worker process before `rmprocs` a good idea? It does seem to work, but throws an ugly stacktrace out to stderr.

tanmaykm replied:
And it was not difficult to avoid that. Added a commit with that now.

tanmaykm committed:
kill worker for shorter test run time

amitmurthy commented:
I just realized that with this PR, now there is a difference in the way that we handle errors during initial master-worker connection setup and errors during the handshake or later (while `addprocs` has still not returned).

Should we just 
1) print warmings for each failed worker startup (under any circumstances) and
2) document that `addprocs` may return less than the requested number of workers?

In any case the fact that `addprocs` may return less than the requested number of workers needs to be documented. While OK going forward, may not be a good idea to backport it?

amitmurthy commented:
Needs docs and further discussions before merging.

tanmaykm commented:
Agree. I would wait a while for conclusion of further opinions/discussions before updating docs.

tanmaykm committed:
document new behavior of addprocs

tanmaykm commented:
Not sure why CI is failing, though tests pass on my local machine. Will investigate this.

amitmurthy commented:
Putting this up for discussion - Shouldn't worker connect or reading host/port errors be converted to warnings and ignored too? i.e., when we cannot cannot connect to some newly launched workers `addprocs` will return a count of the actual workers launched and print warnings instead of throwing errors as it currently does.

For example the errors thrown [here](https://github.com/JuliaLang/julia/blob/df1ee8cb185bbf75999384ce882a89cd9ca8d3fd/stdlib/Distributed/src/cluster.jl#L574-L583) and [here](https://github.com/JuliaLang/julia/blob/df1ee8cb185bbf75999384ce882a89cd9ca8d3fd/stdlib/Distributed/src/cluster.jl#L295-L344) and tested [here](https://github.com/JuliaLang/julia/blob/df1ee8cb185bbf75999384ce882a89cd9ca8d3fd/stdlib/Distributed/test/distributed_exec.jl#L1109-L1172).

tanmaykm commented:
Looks like connect to non-routable IP `203.0.113.0` used to simulate a connect timeout fails immediately instead of blocking. Check `CONNECT FAILED in 0.05166292190551758 secs` in build logs [here](https://build.julialang.org/#/builders/22/builds/4425/steps/2/logs/stdio).

I do not see any way to simulate that condition reliably. So I'll comment that test out for now.

tanmaykm commented:
And yes. Seems to me that we can make some/all of the worker connect errors into warnings now.

tanmaykm committed:
test connect timeout only when JULIA_TESTFULL==1

tanmaykm commented:
CI has passed except one failure in `buildbot/package_win32`, which seems unrelated. Can someone trigger that again?

tanmaykm commented:
CI is passing now. We can take up and discuss https://github.com/JuliaLang/julia/pull/32290#issuecomment-521105039 as a separate PR.

Does this look okay to merge?

tanmaykm commented:
Discovered that `addprocs` can also lock up if `connect` to worker freezes [here](https://github.com/JuliaLang/julia/blob/df1ee8cb185bbf75999384ce882a89cd9ca8d3fd/stdlib/Distributed/src/cluster.jl#L574-L583). Will add more changes to fix that in a bit.

tanmaykm committed:
not fatal to fail to kill worker during deregister

do not throw a fatal error if we could not issue a remote exit to kill the worker while deregistering.

tanmaykm commented:
Also master should not exit with an error if it failed to issue a remote kill to a deregistered worker. Will make changes for that too.

tanmaykm committed:
handle possible timeout during connect to worker

It is possible for a `connect` call from master to worker during worker setup to hang indefinitely. This adds a timeout to handle that, so that master does nto lock up as a result. It simply deregisters and terminates the worker and carries on.

tanmaykm committed:
add tests for connect timeout

vtjnash commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -1516,6 +1516,60 @@ cluster_cookie("foobar") # custom cookie
+end
+@test length(npids) == 0
+@test nprocs() == 1
+@test Distributed.worker_timeout() < t < 360.0
```
Adding up to 6 minutes to the test is probably unacceptable. We should use the existing environment variable to set this timeout to something very small.

tanmaykm replied:
Agree. Pushing a fix in a bit.

tanmaykm replied:
done in https://github.com/JuliaLang/julia/pull/32290/commits/c399338e1ed47041c31c2e01946cbec108139964

tanmaykm committed:
reduce test time for connect timeout test

tanmaykm committed:
rebase changes from master

tanmaykm commented:
Unless there are any further comments, it may be good to have this merged?

ViralBShah commented:
@jeffbezanson can you merge?

ViralBShah commented:
Or perhaps @keno?

Keno commented:
Distributed is extremely unstable currently. I would prefer not to add additional code that until that's properly stablized.

ViralBShah commented:
What is the status of `Distributed` now? Can we merge this?

ViralBShah commented:
Bump

StefanKarpinski commented:
@Keno, since your comment blocked this, please reply (or I'll go ahead and merge it).

vtjnash commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -666,16 +698,16 @@ function create_worker(manager, wconfig)
     wait(rr_ntfy_join)
     if timedout
-        error("worker did not connect within $timeout seconds")
+        @error("worker did not connect within $maxwait seconds")
```
This isn’t really similar, despite the names

StefanKarpinski replied:
Specifically: `@error` logs an error and continues, whereas `error` actually throws.

tanmaykm replied:
Yes, the intention was not to throw an error in this case.

tanmaykm replied:
now removed in https://github.com/JuliaLang/julia/pull/32290/commits/c7ceaea79ab73a9e1f6db7bd7f2dcbada4c4787c

vtjnash commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -603,7 +624,17 @@ function create_worker(manager, wconfig)
+    Timer(1, interval=1) do timer
+        timeout -= 1
+        if timeout <= 0.0
+            put!(rr_ntfy_join, :TIMEDOUT)
```
This seems it could fail

tanmaykm replied:
Yes, should handle it and close timer

tanmaykm replied:
done in https://github.com/JuliaLang/julia/pull/32290/commits/c7ceaea79ab73a9e1f6db7bd7f2dcbada4c4787c

vtjnash commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -572,13 +579,27 @@ end
+        @async try
+            put!(rw_future, connect(manager, w.id, wconfig))
+        catch connect_error
+            put!(rw_future, connect_error)
```
Move put! outside the try/catch, and rethrow on unexpected error types

tanmaykm replied:
The name `connect_error` may be misleading. The intention was to handle any error here in the same manner.

vtjnash commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -666,16 +698,16 @@ function create_worker(manager, wconfig)
     # wait for rr_ntfy_join with timeout
     timedout = false
-    @async (sleep($timeout); timedout = true; put!(rr_ntfy_join, 1))
+    @async (sleep(maxwait); timedout = true; put!(rr_ntfy_join, 1))
```
Keep interpolation of value?

tanmaykm replied:
That doesn't seem necessary, maxwait does not change.

tanmaykm replied:
now removed in https://github.com/JuliaLang/julia/pull/32290/commits/c7ceaea79ab73a9e1f6db7bd7f2dcbada4c4787c

vtjnash commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -592,12 +592,22 @@ It should cause the remote worker specified by `pid` to exit.
+        remote_do(exit, pid) # For TCP based transports this will result in a close of the socket
+                           # at our end, which will result in a cleanup of the worker.
+    catch ex
+        @warn("Failed to kill remote worker $pid. Worker unreachable, unresponsive or already exited.")
```
Discarding ex is probably a bad idea

tanmaykm replied:
Yes, since we have the information, it's a good idea to add it to the warning message.

tanmaykm replied:
done in https://github.com/JuliaLang/julia/pull/32290/commits/cfefd407bbc820d44eaca6b1e98e9474f2fe68f6

vtjnash commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -803,6 +803,7 @@ fetch(v15406)
 remotecall_wait(fetch, id_other, v15406)
 
 # Test various forms of remotecall* invocations
+println("\nTesting various forms of remotecall invocations...")
```
Tests should be silent when successful. Use a testset for groupings.

tanmaykm replied:
Yes. Should be changed all over in this test script.

tanmaykm replied:
done in https://github.com/JuliaLang/julia/pull/32290/commits/4a92628567535dd53881d384706499963aa76f7b. Using testsets for groupings should probably be done in a separate PR as that will result in extensive changes to the test script and unrelated to this PR.

vtjnash commented:
The new tests might be risky (@keno, any guess yourself on them?)—that might involve some of the same code as has been causing the current problems.

ViralBShah commented:
So what needs to be done here?

ViralBShah commented:
@tanmaykm Would you have a chance to get this PR done?

tanmaykm commented:
Yes, will address the review comments.

tanmaykm committed:
show exception along with message when kill fails

- show exception along with message when kill fails
- but do not warn if error is ProcessExitedException for the same process

tanmaykm committed:
add a verbose flag to switch off unnecessary msgs

tanmaykm committed:
close timer on exception to put into rr_ntfy_join

also the additional async task for timeout introduced in #34502 will not be required, because this PR handles that already and also differentiates between timeout and error.

tanmaykm commented:
Added more commits to address review comments. Windows builds have failed, probably needs a rebase with current master.

vtjnash commented:
Bump. We fixed most CI issues, so could you push a rebase and we can try to review this soon

tanmaykm commented:
@vtjnash It is now rebased. Couldn't figure out why windows and macos tests failed though.

vtjnash commented:
It looks like you changed the setup code for that test to remove all workers instead of adding them?

vtjnash commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -1530,6 +1546,64 @@ cluster_cookie("foobar") # custom cookie
+        addprocs_with_testenv(ConnectTimeoutTester())
+    end
+catch ex
+    []
```
should have a test here for `ex`

tanmaykm replied:
done in https://github.com/JuliaLang/julia/pull/32290/commits/60cd4e320fc4547860d9146601f072ba121f5b7b now

vtjnash commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -1530,6 +1546,64 @@ cluster_cookie("foobar") # custom cookie
+end
+@test length(npids) == 0
+@test nprocs() == 1
+@test abs(t - 10.0) <= 5.0
```
```suggestion
@test abs(t - 10.0) <= 25.0
```

tanmaykm replied:
I suppose you suggested this because you suspect the allowed timeout is too tight and we sometimes may exceed it?

tanmaykm committed:
check exception type while testing connect timeout

tanmaykm committed:
ensure we have workers for tests

ViralBShah commented:
@tanmaykm Bumping on this old thread. Is it still ok to merge? A review was requested from @vchuravy

ViralBShah committed:
Update stdlib/Distributed/test/distributed_exec.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

ViralBShah committed:
Merge branch 'master' into tan/par2

ViralBShah commented:
@tanmaykm Any thoughts here? It's an old PR, but if it is still good to go, should we put some effort into getting it merged?

vtjnash commented:
Moved to https://github.com/JuliaLang/Distributed.jl/issues/61
